internal/pkg/skuba/deployments/ssh/kubernetes.go

```diff
@@ -56,24 +58,99 @@ func kubernetesUploadSecrets(errorHandling KubernetesUploadSecretsErrorBehavior)
 	}
 }
 
-func kubernetesInstallNodePattern(t *Target, data interface{}) error {
+func kubernetesParseInterfaceVersions(data interface{}) (string, string, error) {
 	kubernetesBaseOSConfiguration, ok := data.(deployments.KubernetesBaseOSConfiguration)
 	if !ok {
-		return errors.New("couldn't access kubernetes base OS configuration")
+		return "", "", errors.New("couldn't access kubernetes base OS configuration")
 	}
 
 	v, err := version.ParseSemantic(kubernetesBaseOSConfiguration.CurrentVersion)
 	if err != nil {
-		return err
+		return "", "", err
 	}
 	currentVersion := kubernetes.MajorMinorVersion(v)
-	patternName := fmt.Sprintf("patterns-caasp-Node-%s", currentVersion)
+	updatedVersion := ""
 
 	if kubernetesBaseOSConfiguration.UpdatedVersion != "" {
-		updatedVersion := kubernetes.MajorMinorVersion(version.MustParseSemantic(kubernetesBaseOSConfiguration.UpdatedVersion))
-		patternName = fmt.Sprintf("patterns-caasp-Node-%s-%s", currentVersion, updatedVersion)
+		updatedVersion = kubernetes.MajorMinorVersion(version.MustParseSemantic(kubernetesBaseOSConfiguration.UpdatedVersion))
+	}
+	return currentVersion, updatedVersion, nil
+}
+
+func kubernetesFreshInstallAllPkgs(t *Target, data interface{}) error {
+	current, _, err := kubernetesParseInterfaceVersions(data)
+	if err != nil {
+		return err
+	}
+	var pkgs []string
+
+	// Standard packages for a new cluster
+	pkgs = append(pkgs, "+caasp-release", "skuba-update", "supportutils-plugin-suse-caasp", "cri-tools")
+	// Version specific
+	pkgs = append(pkgs, fmt.Sprintf("+kubernetes-%s-kubeadm", current))
+	pkgs = append(pkgs, fmt.Sprintf("+kubernetes-%s-kubelet", current))
+	pkgs = append(pkgs, fmt.Sprintf("+kubernetes-%s-client", current))
+	pkgs = append(pkgs, fmt.Sprintf("+cri-o-%s*", current))
+
+	_, _, err = t.zypperInstall(pkgs...)
+	return err
+}
+
+func kubernetesUpgradeKubeadmPkg(t *Target, data interface{}) error {
+	// zypper install -- -kubernetes-old-kubeadm +kubernetes-new-kubeadm
+	currentV, nextV, err := kubernetesParseInterfaceVersions(data)
+	if err != nil {
+		return err
+	}
+	if nextV == "" {
+		return errors.New("Incorrect upgrade version")
 	}
-	_, _, err = t.ssh("zypper", "--userdata", "skuba", "--non-interactive", "install", patternName)
+
+	var pkgs []string
+
+	if currentV == "1.17" {
+		// on 1.17 we can't remove kubernetes-1.17-kubeadm, doesn't exist.
```

Comment:
Maybe expand the comment and state that k8s v1.17 is the last one installed using the pattern model. Just to provide more context to this "magical" number

Reply:
Yes! Totally.

Reply:
Done.